### ISSUE TITLE
RFC: Rename default branch from "master" to "main"

### ISSUE THREAD

**Issue by stweil** [tesseract-ocr/tesseract#3554]
The new standard for Git and platforms like GitHub and GitLab is using `main` instead of `master` for the default branch:

https://sfconservancy.org/news/2020/jun/23/gitbranchname/
https://github.com/github/renaming
https://about.gitlab.com/releases/2021/06/22/gitlab-14-0-released/#default-branch-name-for-new-repositories-now-main

I'd like to change the tesseract-ocr repositories to that new standard and rename the `master` branch.

This requires updates for existing locally cloned repositories, too (unless they are removed and a new clone is made). Nevertheless I think we should use the standard naming, and switching now is better than switching later.

**Comment by stweil** [comment]
The name change also requires updates of the documentation. It is possible to have both `master` and `main` branches for some time (for example until end of 2021) with `main`being the default branch and `master` no longer updated. That would make the change smoother, for example avoid breaking links in the documentation.

**Comment by amitdo** [comment]
:+1:

**Comment by amitdo** [comment]
>The name change also requires updates of the documentation. It is possible to have both master and main branches for some time (for example until end of 2021) with mainbeing the default branch and master no longer updated. That would make the change smoother, for example avoid breaking links in the documentation.

I don't think it is necessary to keep the old name.

https://github.blog/changelog/2020-07-17-links-to-deleted-branches-now-redirect-to-the-default-branch/

https://github.com/github/renaming#renaming-existing-branches

I suggest to to rename the branch  from the web interface.

**Comment by amitdo** [comment]
Question: When should we do it?

1) This week.
2) Right before the final release of 5.0.0.
3) Right after the final release of 5.0.0.

**Comment by zdenop** [comment]
if to do it, then ASAP, so it relevant docs/links could be adapter before release 5.0.0.

**Comment by stweil** [comment]
Yes, I think it should be done before the next tagging.

**Comment by stweil** [comment]
I tested the rename process for https://github.com/tesseract-ocr/docs. That repository also has GitHub pages https://tesseract-ocr.github.io/docs/. Those have to be activated again after the rename.

**Comment by stweil** [comment]
Now I changed https://github.com/tesseract-ocr/langdata, too. The existing pull requests are updated automatically.

**Comment by stweil** [comment]
All tesseract-ocr repsitories now use `main` as the default branch.

Still open:

- Watch continuous integration whether it still works
- Update more content and links to use "main" instead of "master"

**Comment by bertsky** [comment]
@stweil That's less than 24 hours between raising the issue and enforcing the new rule. Were you interested at all in any actual discussion around this? (You already [had my arguments against this](https://github.com/OCR-D/ocrd_all/pull/176#issuecomment-678632958).)

**Comment by stweil** [comment]
I cite my statement from that discussion (August 2020):

> This is the wrong place to discuss the motivation behind that change. And GitHub suggests to postpone it, so we can wait and decide later this year.

The first sentence still applies. We are now a year later. GitHub no longer suggests to postpone the change. Meanwhile `main` is the new standard. For people who start learning Git and for new repositories that's the norm. Many other Open Source projects already switched to `main`. So for more and more people using the old name will become increasingly strange.

**Comment by bertsky** [comment]
> The first sentence still applies

What a nice way of saying you (still) don't care about the arguments. Where else would you have the discussion then? Just because a company (like Github or SF Conservancy) puts up a certain directive, why should we feel bound by that?

> Meanwhile `main` is the new standard.

For *new projects*, indeed.

> So for more and more people using the old name will become increasingly strange.

I very much doubt that. 

First, there are way more things newcomers (to any project) need to get used to than just the branch names. (While for current contributors and dependent projects, it means extra work in adapting their workflows and deployments.) 

Second, as you know, Open Source has a long tail – most projects have but one or two committers with only sporadic activity. These will unlikely waste their precious time with such cosmetics. But even if everyone abides by corporate terminology: you cannot just disappear that decades long convention, respected by existing documentation, tutorials, FAQs, books.

**Comment by LiteracyFanatic** [comment]
Just thought I'd chime in that this change broke the `tesseract-git` package for Arch Linux. https://aur.archlinux.org/packages/tesseract-git/#news It's not a huge deal or anything, the maintainer just needs to replace `master` with `HEAD` and I can edit the `PKGBUILD` manually to the fix the build myself in the meantime. Arguably it should have been written that way to begin with as a best practice, but the reality of the matter is that for better or for worse lots of existing tooling in the wild *does* assume that the default branch is `master` and changing the default branch of existing repositories *will* break things for people.

To be clear, I'm not demanding the change be undone or anything. Tesseract is wonderful project that I'm very happy to be able to use for free and it's not my place to tell its developers what source control conventions to follow. I just feel that it needs to be pointed out that the cost of this change, however small, is non-zero. Especially when multiplied by the enormous number of projects making such transitions.

**Comment by amitdo** [comment]
Next time we should give more time for feedback about a change with a big impact on our users.

Nevertheless, I still think it's a good change.

**Comment by amitdo** [comment]
Maybe we can re-introduce the `master` branch. It won't be the default but will always follow `main`. Does anyone know how to this automatically ?

**Comment by stweil** [comment]
GitHub already provides backward compatibility, so it is still possible to access `master` and get redirected to `main`.

**Comment by bertsky** [comment]
> GitHub already provides backward compatibility, so it is still possible to access `master` and get redirected to `main`.

Github may do that on their web server, but there is no backwards compatibility on git or http CLI access. That's what will break existing workflows and deployments.

**Comment by stweil** [comment]
> Maybe we can re-introduce the `master` branch. It won't be the default but will always follow `main`.

I had suggested that in my initial proposal. Adding an additional branch named `master` is easy, and synchronisation with `main` could be done automatically.

Nevertheless I think that we should wait and see whether we really need that (and for which of the tesseract-ocr repositories). The change was done for the default branch which is normally used by people doing development. I am sure they can handle the necessary adjustments - if any are required at all.

**Comment by stweil** [comment]
> the maintainer just needs to replace `master` with `HEAD`

We have annotated tags. https://wiki.archlinux.org/title/VCS_package_guidelines#Git suggests better code for that. And that code does not refer to `master`, `main` oder `HEAD`.

**Comment by bertsky** [comment]
> > Maybe we can re-introduce the `master` branch. It won't be the default but will always follow `main`.
> 
> I had suggested that in my initial proposal. Adding an additional branch named `master` is easy, and synchronisation with `main` could be done automatically.
> 
> Nevertheless I think that we should wait and see whether we really need that (and for which of the tesseract-ocr repositories). The change was done for the default branch which is normally used by people doing development. I am sure they can handle the necessary adjustments - if any are required at all.

I find this "let it fail" approach really unfair and unjust, both to external dependent projects and to contributors. It's not hard to imagine what will go wrong without at least a grace period or advance notification.

Example of new CI failures of dependent packages due to rename:
```
--2021-09-14 06:18:08--  https://github.com/tesseract-ocr/tessdata/raw/master/osd.traineddata
Resolving github.com (github.com)... 140.82.113.4
Connecting to github.com (github.com)|140.82.113.4|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2021-09-14 06:18:08 ERROR 404: Not Found.
```

Please add the aliases ASAP!

**Comment by amitdo** [comment]
I re-introduced the master branch for backward compatibility. main is still the default branch. I only did it for this repo.

**Comment by amitdo** [comment]
@stweil commented 3 days ago

> It is possible to have both `master` and `main` branches for some time (for example until end of 2021) with `main`being the default branch and `master` **no longer updated**.

Did you mean that master will be frozen in its current state or at 5.0.0 final release or at EOY?

@amitdo commented 19 hours ago

> Maybe we can re-introduce the `master` branch. It won't be the default but **will always follow** `main`. 

I added my emphasis for the quotes.

**Comment by h1z1** [comment]
> GitHub already provides backward compatibility, so it is still possible to access `master` and get redirected to `main`.

 This change is maddening... You assume this only exists on Github.    Do you seriously not think something like tesseract is not used outside Github by other projects?  WIth their own CI?  And why on Earth was it done so fast?

**Comment by stweil** [comment]
> Please add the aliases ASAP!

Up to now all reported problems in other projects which use Tesseract repositories were caused because something was wrong in those other projects.

So the switch from `master` to `main` had the surprising and for me unexpected side effect of uncovering bugs in other projects.

The main problem is that no project which expects to produce reproducable results must depend on the latest development release of another project because that other project can change (or even break) at any time. Instead it should use well defined releases of other projects (either tagged or identified by their git revision).

>  And why on Earth was it done so fast?

We want to make 5.0.0 soon. And delaying the name switch would not solve any of the problems.

**Comment by stweil** [comment]
> I re-introduced the `master` branch for backward compatibility. `main` is still the default branch. I only did it for this repo.

Please don't do it for more repositories. Fixing dependent projects is not difficult and has to be done anyway (see my previous comment).

I don't think we need that additional master branch and would like to get rid of it again soon, for example when the final 5.0.0 is released. Would that be acceptable?

**Comment by stweil** [comment]
> Example of new CI failures of dependent packages due to rename

That's a bug in OCR-D/ocrd_all (it should have used tagged releases) which is fixed with https://github.com/OCR-D/ocrd_all/pull/275.

**Comment by amitdo** [comment]
>Please don't do it for more repositories.

I didn't have any intention to so. Actually, I only have write perm mission for this repo.

**Comment by zdenop** [comment]
I am sorry for inconvenience and problems with this change. here few of my thoughts on this:

1.	Regardless politics behind this change – “git main” is new standard. I already have repos with “main” branch and it is quiet annoying to try what it right: main or master. So for actively used repos renaming to main is common sense for me.
2.	This change needed to be done before releasing version 5. New release will draw attentions and doing change after main release would cause more problems IMO. That is why I was/am for making this change ASAP/ fast, so we can fix issues on our side soon. We are still in beta phase so significant changes are ok for me.
3.	If somebody uses master/main (without tags) in outside project, there could be only one reasonable explanation for it – to get notified about changes and breaks. Otherwise there is just made wrong setup on their side and they need to fix it. There is no reason why to complain here about broken external project because of changes in not released code.  It is NORMAL to expect that code in master/main is broken or not working as expected. 
4.	In the light of point 2. and 3. creating “master” copy/alias for main repository is WRONG decision IMO. This will hide change - many people will not recognize change and pain of change will last longer. Temporary solutions often become permanent problems.

**Comment by stweil** [comment]
> No, that was [tesserocr](https://github.com/sirfz/tesserocr/runs/3595393850).

Tesserocr ist fixed now: https://github.com/sirfz/tesserocr/commit/fa34eaf853d0802bb1555d46298ef2ee6ede389b.

**Comment by amitdo** [comment]
Okay, you can undo my 'fix'.

**Comment by zdenop** [comment]
amitdo: I do not want to make significant changes each hour. I just send a message to user and devel forums, so users have some time to react.  So let's give them one week or two ;-)

I believe everybody tries to do their best for this project. As a takeaway:  for an important change like this I suggest waiting some time (7 days), so relevant people can react or we can prepare some time schedule, inform people that do not follow the issue tracker.
